src/Items/ItemLilypad.h

```diff
@@ -57,51 +51,146 @@ class cItemLilypadHandler :
 			}
 
 			virtual bool OnNextBlock(int a_CBBlockX, int a_CBBlockY, int a_CBBlockZ, BLOCKTYPE a_CBBlockType, NIBBLETYPE a_CBBlockMeta, eBlockFace a_CBEntryFace) override
-			{
-				if (IsBlockWater(a_CBBlockType))
-				{
-					if ((a_CBBlockMeta != 0) || (a_CBEntryFace == BLOCK_FACE_NONE))  // The hit block should be a source. The FACE_NONE check is clicking whilst submerged
-					{
-						return false;
-					}
-					AddFaceDirection(a_CBBlockX, a_CBBlockY, a_CBBlockZ, BLOCK_FACE_YP);  // Always place pad at top of water block
-					if (
-						!IsBlockWater(a_CBBlockType) &&
-						cBlockInfo::FullyOccupiesVoxel(a_CBBlockType)
-						)
-					{
-						// Can't place lilypad on air / in another block!
-						return true;
-					}
+			{	
+				if((((a_CBBlockType == E_BLOCK_STATIONARY_WATER) || (a_CBBlockType == E_BLOCK_WATER)) && (a_CBBlockMeta == 0))
+				|| (a_CBBlockType == E_BLOCK_ICE) || (a_CBBlockType == E_BLOCK_FROSTED_ICE))
+				{	
 					m_HasHitFluid = true;
 					m_Pos.Set(a_CBBlockX, a_CBBlockY, a_CBBlockZ);
+					m_BlockFace = a_CBEntryFace;
 					return true;
+				
 				}
 				return false;
 			}
 
 			Vector3i m_Pos;
+			eBlockFace m_BlockFace;
 			bool m_HasHitFluid;
-
 		};
 
 		cCallbacks Callbacks;
-		cLineBlockTracer Tracer(*a_Player->GetWorld(), Callbacks);
+		cLineBlockTracer Tracer(*a_World, Callbacks);
 		Vector3d Start(a_Player->GetEyePosition() + a_Player->GetLookVector());
 		Vector3d End(a_Player->GetEyePosition() + a_Player->GetLookVector() * 5);
 
 		Tracer.Trace(Start.x, Start.y, Start.z, End.x, End.y, End.z);
 
-		if (Callbacks.m_HasHitFluid)
+		if (!Callbacks.m_HasHitFluid)
+		{
+			return false;
+		}
+		
+		// Since we hit fluid, try placing the block on top of the fluid we found
+		// Mimick clicking on top of the fluid to do this
+		return OnPlayerPlace(*a_World, *a_Player, a_Item,
+			Callbacks.m_Pos.x, Callbacks.m_Pos.y, Callbacks.m_Pos.z, Callbacks.m_BlockFace,
+			0, 0, 0
+		);
+	}
```

Comment:
The entire `OnItemUse` function is now a big ball of wire with only too-detailed-comments sprinkled in it, with no overview. I assume it works by first tracing the line-of-sight, then if a water block is hit, relaying to `OnPlayerPlace`. If so, split the line-of-sight tracing into a separate function so that the `OnItemUse` is kept short and concise.